### PR TITLE
SAMR21: Optimizations to make RPL_UDP example works

### PR DESCRIPTION
**Pull request by maxencechotard**
Optimizations to make samr21-xpro board works with RPL_UDP example :  
-Setting channel to 12 between  AT86RF231_MIN_CHANNEL (11) to AT86RF231_MAX_CHANNEL (26).
-Modification of makefile to allow RPL_UDP example to work on samr21-xpro board.
-Modification of radio and rpl_process thread size to optimize bss.
-Modification of RPL_MAX_ROUTING_ENTRIES to fit RAM size.

### CODE REVIEW

**Comment by gebart** [comment]
(kicked Travis.)

See also #2325

File: examples/rpl_udp/rpl_udp.h [hunk 1]
```diff
@@ -15,7 +15,11 @@ extern "C" {
 
 #define APP_VERSION "1.2"
 
+#ifdef MODULE_AT86RF231
+#define RADIO_CHANNEL   (12)
+#else
 #define RADIO_CHANNEL   (10)
+#endif
```
Comment (gebart) [inline examples/rpl_udp/rpl_udp.h]: I believe this will be unnecessary after #2325 is merged.

**Comment by gebart** [comment]
Original issue: #2173

File: boards/samr21-xpro/include/board.h [hunk 2]
```diff
@@ -48,6 +48,13 @@ extern "C" {
 #define AT86RF231_SLEEP    GPIO_7
 
 #define AT86RF231_SPI_SPEED SPI_SPEED_1MHZ
+
+/*Thread optimization for samr21*/
+#define RADIO_STACK_SIZE            (1024)
+#define RPL_PROCESS_STACKSIZE           (1024)
```
Comment (cgundogan) [inline boards/samr21-xpro/include/board.h]: may look better with same indentations

File: boards/samr21-xpro/include/board.h [hunk 3]
```diff
@@ -48,6 +48,13 @@ extern "C" {
 #define AT86RF231_SLEEP    GPIO_7
 
 #define AT86RF231_SPI_SPEED SPI_SPEED_1MHZ
+
+/*Thread optimization for samr21*/
+#define RADIO_STACK_SIZE                (1024)
+#define RPL_PROCESS_STACKSIZE           (1024)
+
+/*samr21 is very memory constrained and does not run with the default RPL_MAX_ROUTING_ENTRIES (= 128)*/
+#define RPL_MAX_ROUTING_ENTRIES            (8)
```
Comment (cgundogan) [inline boards/samr21-xpro/include/board.h]: will this line interfere with the `RPL_MAX_ROUTING_ENTRIES` specified by command line (at compilation time) ? Which one has the higher priority?

Comment (BytesGalore) [inline boards/samr21-xpro/include/board.h]: Yes, the one in this header is defined after the command line option, resulting in:
```
./RIOT/boards/samr21-xpro/include/board.h:57:0: warning: "RPL_MAX_ROUTING_ENTRIES" redefined
 #define RPL_MAX_ROUTING_ENTRIES            (8)
```
so I vote to change it to:
```
#ifndef RPL_MAX_ROUTING_ENTRIES
#   define RPL_MAX_ROUTING_ENTRIES            (8)
#endif
```

**Comment by maxencechotard** [comment]
@BytesGalore I have to try if setting RPL_MAX_ROUTING_ENTRIES to something higher than 8 still make RPL_UDP example works on samr21

**Comment by BytesGalore** [comment]
@maxencechotard indeed you can probably force a maximum with:
```
#ifndef RPL_MAX_ROUTING_ENTRIES
#   define RPL_MAX_ROUTING_ENTRIES            (8)
#else
#if (RPL_MAX_ROUTING_ENTRIES > 8)
#   undef RPL_MAX_ROUTING_ENTRIES
#   define RPL_MAX_ROUTING_ENTRIES            (8)
#endif
#endif
```

**Comment by cgundogan** [comment]
I am not sure if C macros have lazy evaluation for logical operators, but you could also try the following, to have it in a slightly more minimal way:
```
#if !defined(RPL_MAX_ROUTING_ENTRIES) || RPL_MAX_ROUTING_ENTRIES > 8
#    undef RPL_MAX_ROUTING_ENTRIES
#    define RPL_MAX_ROUTING_ENTRIES (8)
#endif
```

Anyway, probably the size of the routing table should be printed somewhen during compilation or when the application starts / rpl initializes, to see if my preferred size (lets say >8) was overridden with 8 or not.
@BytesGalore  what is your opinion on that?

**Comment by BytesGalore** [comment]
@cgundogan 
+1 to slim the `RPL_MAX_ROUTING_ENTRIES` handling
+1 to print the set routing table size on compilation (maybe only if the desired size has been trimmed)

**Comment by BytesGalore** [comment]
@maxencechotard youre totally investigating the `ISR_HARD_FAULT` from the lower bound has priority.
Btw. I have some numbers for the BSS sizes from my "tests":
```
RPL_MAX_ROUTING_ENTRIES=4:
  text	   data	    bss	    dec	    hex	filename
 137300	    244	  31112	 168656	  292d0	...samr21-xpro/rpl_udp.elf

RPL_MAX_ROUTING_ENTRIES=8:
   text	   data	    bss	    dec	    hex	filename
 137300	    244	  31256	 168800	  29360	...samr21-xpro/rpl_udp.elf

RPL_MAX_ROUTING_ENTRIES=16:
   text	   data	    bss	    dec	    hex	filename
 137300	    244	  31544	 169088	  29480	...samr21-xpro/rpl_udp.elf

RPL_MAX_ROUTING_ENTRIES=24:
  text	   data	    bss	    dec	    hex	filename
 137300	    244	  32120	 169664	  296c0	...samr21-xpro/rpl_udp.elf

RPL_MAX_ROUTING_ENTRIES=31:
  text	   data	    bss	    dec	    hex	filename
 137300	    244	  32080	 169624	  29698	...samr21-xpro/rpl_udp.elf

RPL_MAX_ROUTING_ENTRIES=32:
   text	   data	    bss	    dec	    hex	filename
 137300	    244	  32120	 169664	  296c0	...samr21-xpro/rpl_udp.elf
```
GCC:
`arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 4.9.3 20141119 (release) [ARM/embedded-4_9-branch revision 218278]`

**Comment by thomaseichinger** [comment]
Did you manage to run the UDP part of the `rpl_udp` example? Gave it a try and it crashes either when sending UDP messages (the sending node) or sometimes the receiving node.

File: sys/net/include/rpl.h [hunk 4]
```diff
@@ -41,7 +41,9 @@ extern "C" {
 #define CC1100_RADIO_MODE CC1100_MODE_WOR
 
 #define RPL_PKT_RECV_BUF_SIZE 16
+#ifndef RPL_PROCESS_STACKSIZE
 #define RPL_PROCESS_STACKSIZE KERNEL_CONF_STACKSIZE_MAIN
+#endif
```
Comment (LudwigOrtmann) [inline sys/net/include/rpl.h]: AFAIK It's just sloppy style to not have the `ifndef`s.
Also, yes, do a separate PR for these please.

**Comment by maxencechotard** [comment]
@thomaseichinger Sorry for my late answer. I actually already made UDP  works without the RPL part of RPL_UDP example. During my tries it does not crashed and still works pretty well. This my working branch on github : <code>https://github.com/maxencechotard/RIOT/tree/UDP_ONLY</code>

**Comment by bapclenet** [comment]
Let's close this PR as well as it's not going to make any sense with the new network stack and the new at86rf2xx driver?
@gebart

**Comment by gebart** [comment]
@bapclenet I tend to agree, since this PR has been quite inactive lately.

Has the new network stack been tested with the SAMR21?

**Comment by bapclenet** [comment]
I've seen that UDP has been merged recently, I'm going to try soon. Concerning, RPL, the implementation with new Network Stack is not done yet.

**Comment by bapclenet** [comment]
I close this PR since ng_udp has been merged and RPL is on the go.
